Source/missiles.h

```diff
@@ -97,7 +97,8 @@ struct Missile {
 	MissilePosition position;
 	int _mimfnum; // The direction of the missile (direction enum)
 	int _mispllvl;
-	bool _miDelFlag; // Indicate whether the missile should be deleted
+	bool _miDelFlag;     // Indicate whether the missile should be deleted
+	bool _miSpellFizzle; // Indicate whether a spell-spawned missile should consume mana

```

Comment:
`_miSpellFizzle` is only used in `AddGuardian/AddBersersk/...` and `CastSpell` or?

`CastSpell` -> `AddMissile` -> `AddGuardian` and when `AddMissile` returns `CastSpell` checks `_miSpellFizzle`.

We could avoid this member in `Missile` and instead add it to `AddMissile` and `AddMissileParameter`.
`AddMissile` would need to return spell fizzle (for example a new function overload).
`AddMissile` would get spell fizzle from `AddMissileParameter` (there it is added as member).
`AddGuardian/AddBersersk/...` could set this new member in `AddMissileParameter`.

What do you think? 🙂

Reply:
It is true that the field is only used by `CastSpell()` to determine whether the spell cost should be applied. However, I would also argue that there should be other contexts in which it might be useful to know whether a spell fizzled, like maybe a fallback or retry with different constraints. It's possible that no such mechanics were developed for Diablo or Hellfire because Blizzard didn't develop a mechanism like this one for returning that information to the caller.

I wasn't sure about putting it in `AddMissileParameter` because that gets passed by const reference. In order to return the value of spell fizzle via `AddMissileParameter`, we'd have to pass by mutable reference, which isn't a big deal except maybe semantically. The const reference suggests that `AddMissileParameter` was intended for passing information into the `AddMissile()` function rather than for getting information out.

That said, I know that the const qualifier is sometimes applied simply because there is currently no use-case for mutability. Naturally, we can change the semantics if we need to, but since the `AddMissile()` function already returns a reference to the missile and `missile->_miSpellFizzle` makes some good sense semantically, this is ultimately what I settled on. If there's a compelling reason to change it then I could certainly be convinced, but a function overload with a different return value does sound considerably less intuitive to me.

Reply:
I added `AddMissileParameter` in #3957 and at that time I used `const` cause there was no use-case for mutability. 😉

I suggested the change, cause I thought about save game compatibility (the new flag is not part of the save game). Then I saw every time we set `_miSpellFizzle` we also set `_miDelFlag`. That means the missile don't life very long and I thought this new flag is more part of the missile creation then part of the missile behavior.
But yeah a new overload is also not very nice and intuitive.

Alternative idea:
Currently `AddMissile` only returns `nullptr` when we have more missiles then `size_t` allows (this is a theoretically limit. before that we would run out of memory).
Could we change `AddMissile` to return a `nullptr` if the missile creation fails / fizzle?
That way we don't need a new overload and `AddMIssile` would make more sense.
We could change the `AddGuardian/AddBersersk/...` to return a bool value (or alternative change `AddMissileParameter`).

🤔

Reply:
That alternative sounds pretty good to me. I guess I'd be more inclined to change `AddMissileParameter` mutability than return a bool since most `AddMissile()` functions don't have conditions in which they could fizzle.

Reply:
I had the same thought (that's why I stated the alternative).
But I was unsure what's more ideal.
Adding a return bool would be cleaner in some way "creation failed or not".
On the other hand using `AddMissileParameter` we would only need to change methods that are affected by it. And I'm unsure if a return bool is really more intuitive, cause we have some missiles that are directly deleted (for example golem or identify. In this case a return false would also make sense in a way (cause we don't need a `Missile` object). That means a fizzle member in `AddMissileParameter` would make it more cleaner. 🤔 
That's why I'm also in favor of using `AddMissileParameter` (but not a strong opinion).

Reply:
> And I'm unsure if a return bool is really more intuitive, cause we have some missiles that are directly deleted (for example golem or identify. In this case a return false would also make sense in a way (cause we don't need a `Missile` object).

Yeah, I 100% agree that the bool return value could be ambiguous. The reason I find the `nullptr` return value to be more acceptable is because the intent is clear from reading the code in the `AddMissile()` function. If we had every specialization return a bool instead, then the meaning of that bool would only be apparent when reading the code of certain functions and not others. This could perhaps be mitigated with comments, but I think the `AddMissileParameter` approach is more effective.

Anyway, I made the change so feel free to review and see how you like it.

Reply:
I looked at the code and I think it's good. 🙂 
Even if we want to change it later, it should be fairly easy.